### PR TITLE
CLOUDP-36271 integrate oplog tailing and writing to S3

### PR DESCRIPTION
**Pull request by chunmingli**
https://jira.mongodb.org/browse/CLOUDP-36271

I introduced a few stages (op_reader -> slicer(slice maker) -> slice pusher) similar to the pipelining idea mentioned in https://blog.golang.org/pipelines, where each stage consumes from a channel that the previous stage produces and produce something for the next stage to consume via a channel. Each stage is responsible to close the channel it created.

Integrating with the config will come in a follow up ticket

### CODE REVIEW

File: cps-pit/tail.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	go func() {
+		time.Sleep(3 * time.Second)
+		cancel()
+		errc <- ctx.Err()
+		time.Sleep(3 * time.Second)
+	}()
+	//*/
+
+	<-errc
+	cancel()
```
Comment (lawctan) [inline cps-pit/tail.go]: defer cancel instead early on

File: cps-pit/op_reader.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		raw, err := getDoc(ctx, cur)
+		if err != nil {
+			println("error in produceDocs", err)
+			errc <- err
+			return
+		}
+		select {
+		case <-ctx.Done():
+			println("done signaled, closing docs")
+			close(docs)
```
Comment (lawctan) [inline cps-pit/op_reader.go]: move close(docs) to beginning of function and add `defer`

File: cps-pit/slicer.go [hunk 3]
```diff
@@ -50,66 +50,84 @@ func NewSlicer(
...
-		return false
+		if err != nil {
+			println("error in produceSlices", err)
+			slicer.errc <- err
+			return
+		}
+		select {
+		case <-slicer.ctx.Done():
+			println("done signaled, closing slices")
+			close(slices)
```
Comment (lawctan) [inline cps-pit/slicer.go]: move close(slices) to beginning of function and add `defer`

File: cps-pit/tail.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	/* the following example isto have a separate monitoring goroutine that cancels all operations
+	go func() {
+		time.Sleep(3 * time.Second)
+		cancel()
+		errc <- ctx.Err()
+		time.Sleep(3 * time.Second)
+	}()
+	//*/
+
+	<-errc
```
Comment (lawctan) [inline cps-pit/tail.go]: would this block if there are no errors?

Comment (chunmingli) [reply cps-pit/tail.go]: Correct, it would block. Otherwise the "main thread" will exit regardless of all spawned pipelining goroutines.

Comment (lawctan) [reply cps-pit/tail.go]: Ah I see, I get it now. That's an interesting way to block.

File: cps-pit/tail.go (hunk 1 above)
Comment (chunmingli) [reply cps-pit/tail.go]: Cool - eventually we'll have another goroutine (something like what's commented out L61-68) to decide when to cancel.

**Pull request merged**